subsys/usb/class/netusb/function_ecm.c

```diff
@@ -345,21 +341,46 @@ static struct netusb_function ecm_function = {
 	.send_pkt = ecm_send,
 };
 
-static inline void ecm_status_interface(const uint8_t *desc)
+static int ecm_custom_handler(struct usb_setup_packet *setup, int32_t *len,
+			      uint8_t **data)
 {
-	const struct usb_if_descriptor *if_desc = (void *)desc;
-	uint8_t iface_num = if_desc->bInterfaceNumber;
-	uint8_t alt_set = if_desc->bAlternateSetting;
+	uint8_t iface = (setup->wIndex) & 0xFF;
+	uint8_t alt = (setup->wValue) & 0xFF;
 
-	LOG_DBG("iface %u alt_set %u", iface_num, if_desc->bAlternateSetting);
+	/* Second interface for ecm function has alternate.
+	 * Check if the request addressed this interface, if no
+	 * immediately return as this request is not directed
+	 * to this function.
+	 */
+	if (iface != ecm_if_data[1]->bInterfaceNumber) {
+		return -ENOTSUP;
+	}
 
-	/* First interface is CDC Comm interface */
-	if (iface_num != ecm_get_first_iface_number() + 1 || !alt_set) {
-		LOG_DBG("Skip iface_num %u alt_set %u", iface_num, alt_set);
-		return;
+	if (REQTYPE_GET_RECIP(setup->bmRequestType) ==
+	    REQTYPE_RECIP_INTERFACE) {
+		switch (setup->bRequest) {
+		case REQ_SET_INTERFACE:
+			/* Check if proper alternate setting is requested */
+			if (alt == ecm_if_data[1]->bAlternateSetting) {
+				alternate_setting = alt;
+			} else if (alt == ecm_if_data[2]->bAlternateSetting) {
+				alternate_setting = alt;
+				netusb_enable(&ecm_function);
+			} else {
+				return -ENOTSUP;
+			}
+			LOG_DBG("Set interface iface: %d, alt: %d", iface, alt);
+			return -EINVAL;
```

Comment:
Yes it was but, ekhmm its Zephyr. Let me explain why. We have recently introduced this behaviour in audio PR. `custom_handler` has its special [return type](https://github.com/zephyrproject-rtos/zephyr/blob/master/include/usb/usb_device.h#L160). This is done like that because class implementation handles the status of alternate interface and for set/get request it saves/returns the number of alternate that was requested. For SET however one more thing needs to be done -> proper endpoints associated with that alternate must be enabled or disabled.

Class custom handler is called from [there](https://github.com/zephyrproject-rtos/zephyr/blob/master/subsys/usb/usb_device.c#L915). Look closely at this if condition. If I `return 0` there it will return from `usb_handle_standard_request` without calling `usb_handle_std_interface_req` which enabled/disables the associated endpoints.

Reference: [audio](https://github.com/zephyrproject-rtos/zephyr/blob/master/subsys/usb/class/audio/audio.c#L736).

Reply:
A comment like "This request needs further processing in the core USB stack, hence -EINVAL is returned." would be helpful then. In the audio case, too.

Reply:
I will add this as a separate commit.

Reply:
This behavior is confusing, we have to rework it somehow asap. What about not to do it in custom_handler but implement a specific handler to set interface which would be called from REQ_SET_INTERFACE(usb_set_interface)?